Add missing operations

jd-lara opened the pull request:
Fixes #40 
Fixes #43

jd-lara committed:
add operation for +(param, GAEp)

jd-lara committed:
add convert for GAEp -> PAE

jd-lara committed:
add add_to_expression! functions

jd-lara committed:
add zero

jd-lara committed:
add  +(::GAEp, Param)

jd-lara committed:
add conversion GAEp -> PAE

jd-lara committed:
add copy(PAE)

jd-lara committed:
fix typo

jd-lara committed:
add one operator

joaquimg commented on src/operators.jl:
```diff
@@ -106,6 +107,12 @@ Base.:(-)(lhs::GAEp{C}, rhs::PAE{C}) where {C} = PAE{C}(-rhs.v,lhs-rhs.p)
 =#
 
+# Make zero
+Base.zero(::Type{PAE{C}}) where {C} = GAEp{C}(zero(C))
```
We could return a PAE here

jd-lara replied:
solved via a1b2761

joaquimg commented on src/operators.jl:
```diff
@@ -106,6 +107,12 @@ Base.:(-)(lhs::GAEp{C}, rhs::PAE{C}) where {C} = PAE{C}(-rhs.v,lhs-rhs.p)
+Base.zero(::Type{PAE{C}}) where {C} = GAEp{C}(zero(C))
+
+# Make one
+Base.one(::Type{PAE{C}}) where {C} = GAEp{C}(one(C))
```
same here

jd-lara replied:
solved via a1b2761

joaquimg commented on src/operators.jl:
```diff
@@ -45,8 +45,8 @@ Base.:(+)(lhs::Parameter, rhs::GAEv{C}) where {C} = PAE{C}(copy(rhs),GAEp{C}(zer
-Base.:(+)(lhs::Parameter, rhs::PAE{C}) where {C} = PAE{C}(copy(rhs.v),lhs+rhs.p)
-Base.:(-)(lhs::Parameter, rhs::PAE{C}) where {C} = PAE{C}(-rhs.v,lhs-rhs.p)
+Base.:(+)(lhs::Parameter, rhs::PAE{C}) where {C} = PAE{C}(copy(rhs.v),JuMP.add_to_expression!(rhs.p,lhs))
+Base.:(-)(lhs::Parameter, rhs::PAE{C}) where {C} = PAE{C}(-rhs.v,JuMP.add_to_expression!(-1*rhs.p,lhs))
```
We have to be careful here, `JuMP.add_to_expression!` updates `expression` *in place*, but the basic operations cannot destroy the original input. This is certainly happening in `+` and might be happening in `-`. Thats the reason for the previous implementation.

jd-lara replied:
I understand the problem here. 

Then we need to implemented properly `+(::Parameter, ::GenericAffExpr{Float64,Parameter})` because in b6116a9 what I did was calling `JuMP.add_to_expression!` and it will create the same update in place. 

Any ideas on what's the best avenue?

jd-lara replied:
I implemented `+(::Parameter, ::GenericAffExpr{Float64,Parameter})` and `-(::Parameter, ::GenericAffExpr{Float64,Parameter})` in the same way at it is done un JuMP

https://github.com/JuliaOpt/JuMP.jl/blob/7426454d1271e78a088b7fce1cdf0374cf6d6dd1/src/operators.jl#L131

See e3b4334

joaquimg reviewed (changes requested):
looks good in general, but needs some updates

jd-lara committed:
change return of one and zero

jd-lara committed:
update operators

jd-lara committed:
fix zero and one

jd-lara committed:
add zero and one for GAEp

jd-lara committed:
add zero() and one() operators over GAEp

jd-lara committed:
zeros and ones needed for JuMP @ operations

jd-lara committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

enable param + GAEp

# The commit message #2 will be skipped:

# enable param + GAEp

jd-lara committed:
enable param + GAEp

jd-lara committed:
enable param + GAEp

jd-lara committed:
Fix Copy

jd-lara committed:
add iszero

jd-lara commented on src/mutable_arithmetics.jl:
```diff
@@ -80,9 +80,32 @@ end
+
+#= 
+
+Convert
```
@joaquimg these convert methods are still missing in #45 and needed in PowerSimulations.

joaquimg replied:
Okay!
I will add the first for sure.
are you sure the second is needed?
a GAEp should never be generated, an manipulated externaly, so it might be bug.
Unless some of the ParameterJuMP functions need it.

joaquimg replied:
is there any example? or its to hard to create one?

joaquimg commented:
replaced by #45
